src/main/java/org/galatea/starter/service/IexService.java

```diff
@@ -48,38 +56,192 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
     }
   }
 
-  /** Get historical prices for the symbol, range, and date combination
-   * @param symbol a valid symbol
-   * @param range an optional range to check
-   * @param date an optional date to check; has priority over range
+  /** Get historical prices for the symbol, range, and date combination.
+   * If a range is provided, get a list of dates in that range and
+   * query them independently to build a result of responses.
+   * Use stored results when available. If there are more than one instance
+   * of historical prices for a date, use the most recent one.
+   * Store values for quicker retrieval.
+   * @param symbol a symbol to reference
+   * @param range a range to check prices for
+   * @param date a date to check prices for; has priority to range if both are included
    */
-  public List<IexHistoricalPrices> getHistoricalPricesForSymbols(final String symbol, final String range, final String date) throws Exception{
+  public List<IexHistoricalPrice> getHistoricalPricesForSymbols(
```

Comment:
This whole method is wayyy too long and has a lot of duplicated logic in it. Can you refactor so that
1. There is no duplication, extract duplicate lines of code to re-usable private methods
2. There are no methods longer than 5 lines long
3. There are no nested blocks (if, for, etc) deeper than two levels (.e.g for { if { is okay, for { if { if is not) (further private methods can be used to remove nesting)